commands/command_filter_process.go

```diff
@@ -0,0 +1,128 @@
+package commands
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/github/git-lfs/git"
+	"github.com/github/git-lfs/lfs"
+	"github.com/spf13/cobra"
+)
+
+const (
+	// cleanFilterBufferCapacity is the desired capacity of the
+	// `*git.PacketWriter`'s internal buffer when the filter protocol
+	// dictates the "clean" command. 512 bytes is (in most cases) enough to
+	// hold an entire LFS pointer in memory.
+	cleanFilterBufferCapacity = 512
+
+	// smudgeFilterBufferCapacity is the desired capacity of the
+	// `*git.PacketWriter`'s internal buffer when the filter protocol
+	// dictates the "smudge" command.
+	smudgeFilterBufferCapacity = git.MaxPacketLength
+)
+
+// filterSmudgeSkip is a command-line flag owned by the `filter-process` command
+// dictating whether or not to skip the smudging process, leaving pointers as-is
+// in the working tree.
+var filterSmudgeSkip bool
+
+// filterSmudge is a gateway to the `smudge()` function and serves to bail out
+// immediately if the pointer decoded from "from" has no data (i.e., is empty).
+// This function, unlike the implementation found in the legacy smudge command,
+// only combines the `io.Reader`s when necessary, since the implementation
+// found in `*git.PacketReader` blocks while waiting for the following packet.
+func filterSmudge(from io.Reader, to io.Writer, filename string) error {
+	var pbuf bytes.Buffer
+	from = io.TeeReader(from, &pbuf)
+
+	ptr, err := lfs.DecodePointer(from)
+	if err != nil {
+		// If we tried to decode a pointer out of the data given to us,
+		// and the file was _empty_, write out an empty file in
+		// response. This occurs because when the clean filter
+		// encounters an empty file, and writes out an empty file,
+		// instead of a pointer.
+		//
+		// TODO(taylor): figure out if there is more data on the reader,
+		// and buffer that as well.
```

Comment:
After discussing in meat-space, the consensus was there will be no more data on the io.Reader coming from Git, since we will have decoded the entire pointer (or at least, tried to) by this point, so I went ahead and removed the TODO.

@larsxschneider can clarify anything that I left out.